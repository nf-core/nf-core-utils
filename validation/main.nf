#!/usr/bin/env nextflow

// Import plugin functions
include { 
    sayHello;
    getWorkflowVersion;
    checkCondaChannels;
    completionSummary;
    generateVersionReport;
    generateCitationReport;
    processMixedVersionSources;
    processMixedCitationSources;
} from 'plugin/nf-core-utils'

workflow {
    
    log.info """
    ================================
    nf-utils Plugin Integration Test
    ================================
    """
    
    // Test 1: Basic sayHello function
    log.info "Testing sayHello function..."
    sayHello("nf-utils integration test")
    
    // Test 2: Get workflow version
    log.info "Testing getWorkflowVersion function..."
    version = getWorkflowVersion()
    log.info "Workflow version: ${version}"
    
    // Test 3: Check conda channels
    log.info "Testing checkCondaChannels function..."
    conda_check = checkCondaChannels()
    log.info "Conda channels check: ${conda_check}"
    
    // Test 4: Test version processing with sample data
    log.info "Testing version processing..."
    
    // Sample topic channel data: [process, name, version]
    sample_versions = [
        ['FASTQC', 'fastqc', '0.11.9'],
        ['MULTIQC', 'multiqc', '1.12'],
        ['TRIMGALORE', 'trimgalore', '0.6.7'],
        ['TRIMGALORE', 'cutadapt', '4.1']
    ]
    
    // Test version report generation
    version_report = generateVersionReport(sample_versions, [])
    log.info "Version report generated successfully"
    
    // Test 5: Test citation processing with sample data
    log.info "Testing citation processing..."
    
    // Create sample meta.yml files for testing
    sample_meta_paths = []
    
    // Test mixed version sources (topic + legacy)
    log.info "Testing mixed version sources..."
    mixed_versions = processMixedVersionSources(sample_versions, [])
    log.info "Mixed versions processed successfully"
    
    // Test 6: Completion summary
    log.info "Testing completion summary..."
    completionSummary(false)
    
    log.info """
    ================================
    All integration tests completed!
    ================================
    """
}
