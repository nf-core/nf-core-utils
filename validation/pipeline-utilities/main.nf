#!/usr/bin/env nextflow

/*
 * E2E validation test for nf-core-utils plugin pipeline utilities
 *
 * This test validates the core pipeline utility functions:
 * - getWorkflowVersion(): Generate workflow version strings
 * - dumpParametersToJSON(): Save pipeline parameters to JSON file
 */

// Import plugin functions for pipeline utilities
include { getWorkflowVersion   } from 'plugin/nf-core-utils'
include { dumpParametersToJSON } from 'plugin/nf-core-utils'

workflow {

    log.info("==========================================")
    log.info("Pipeline Utilities Validation Test")
    log.info("==========================================")

    // Test 1: getWorkflowVersion() function
    log.info("=== Testing getWorkflowVersion() function ===")

    // Test with default parameters (using session info)
    def workflow_version_default = getWorkflowVersion()
    log.info("Default workflow version: ${workflow_version_default}")

    // Test with explicit version
    def workflow_version_explicit = getWorkflowVersion("2.1.0")
    log.info("Explicit version (2.1.0): ${workflow_version_explicit}")

    // Test with version and commit ID
    def workflow_version_with_commit = getWorkflowVersion("2.1.0", "abc123def456789")
    log.info("Version with commit (2.1.0, abc123): ${workflow_version_with_commit}")

    // Test edge cases
    def workflow_version_v_prefix = getWorkflowVersion("v3.0.0")
    log.info("Version with v prefix (v3.0.0): ${workflow_version_v_prefix}")

    def workflow_version_null_version = getWorkflowVersion(null, "xyz789abcdef123")
    log.info("Null version with commit: ${workflow_version_null_version}")

    // Test 2: dumpParametersToJSON() function
    log.info("=== Testing dumpParametersToJSON() function ===")

    // Create simple test parameters
    def test_params = [
        input: 'SRR_Acc_List.txt',
        email: 'test@example.com',
        outdir: './results',
        genome: 'GRCh38',
        skip_fastqc: false,
        max_memory: '128.GB',
        max_cpus: 16,
    ]

    // Test parameter dumping to JSON
    def output_dir = "${workflow.workDir}/validation_output"
    log.info("Dumping parameters to JSON in: ${output_dir}")

    try {
        dumpParametersToJSON(output_dir, test_params)
        log.info("âœ… Parameters dumped successfully")
    }
    catch (Exception e) {
        log.error("Error dumping parameters: ${e.message}")
        log.error("Error details: ${e}")
    }

    // Test edge cases
    log.info("=== Testing edge cases ===")

    // Test with null output directory
    try {
        dumpParametersToJSON(null, test_params)
        log.info("âœ… Null output directory handled correctly")
    }
    catch (Exception e) {
        log.error("Error with null output: ${e.message}")
    }

    // Validate outputs
    log.info("=== Validating outputs ===")

    // Check if JSON file was created
    def json_file_path = "${output_dir}/params.json"
    def json_file = new File(json_file_path)

    if (json_file.exists()) {
        log.info("âœ… JSON file created: ${json_file_path}")

        // Validate JSON content
        try {
            def json_content = new groovy.json.JsonSlurper().parseText(json_file.text)
            log.info("âœ… JSON content is valid")
            log.info("JSON contains ${json_content.size()} parameters")

            // Basic content validation
            assert json_content.input == 'SRR_Acc_List.txt'
            assert json_content.max_cpus == 16
            log.info("âœ… JSON content validation passed")
        }
        catch (Exception e) {
            log.error("JSON validation failed: ${e.message}")
        }
    }
    else {
        log.warn("JSON file not found at: ${json_file_path}")
    }

    // Validate version strings
    log.info("=== Validating version strings ===")

    assert workflow_version_explicit.startsWith('v2.1.0')
    assert workflow_version_with_commit.contains('-g')
    assert workflow_version_v_prefix == 'v3.0.0'

    log.info("âœ… All version string validations passed")

    log.info("==========================================")
    log.info("Pipeline Utilities Validation Complete")
    log.info("==========================================")

    workflow.onComplete {
        log.info(
            """
            ==========================================
            Pipeline Utilities Validation Results
            ==========================================

            âœ… getWorkflowVersion() function tested successfully
            âœ… dumpParametersToJSON() function tested successfully
            âœ… Edge cases handled properly
            âœ… Output validation completed

            Functions ready for use in fetchngs pipeline! ðŸš€
            ==========================================
            """.stripIndent().trim()
        )
    }
}
