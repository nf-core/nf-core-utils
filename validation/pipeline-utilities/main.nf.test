nextflow_pipeline {

    name "Test nf-core-utils pipeline utilities"
    script "../pipeline-utilities/main.nf"
    config "../pipeline-utilities/nextflow.config"

    test("Pipeline utilities") {

        then {
            assert workflow.success
            assert snapshot(
                filterNextflowOutput(workflow.stdout + workflow.stderr, keepAnsi: true)
            ).match()
        }
    }
}
