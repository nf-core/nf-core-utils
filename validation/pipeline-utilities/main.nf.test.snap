{
    "pipeline_utilities_success": {
        "content": [
            [
                "Pipeline is starting! \ud83d\ude80",
                "Pipeline complete! \ud83d\udc4b"
            ],
            0,
            true
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-08-22T11:58:57.522014"
    },
    "utilities_validation_success": {
        "content": [
            true,
            0
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-08-22T11:59:03.385702"
    },
    "utility_testing_complete": {
        "content": [
            true,
            0
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-08-22T11:59:00.426522"
    }
}